src/lazy/streaming_raw_reader.rs

```diff
@@ -345,6 +362,64 @@ impl<SliceType: AsRef<[u8]>> IoBufferHandle for IonSlice<SliceType> {
         let local_end = bytes.len();
         IoBuffer::new(stream_position, bytes, local_offset, local_end)
     }
+
+    #[cfg(feature = "lazy-source-location")]
+    fn row(&self) -> usize {
+        // Update row for location metadata
+        let available_range = ..self.prev_position;
+        let data  = &self.stream_bytes()[available_range];
+        if !data.is_empty() {
+            // Calculate `rows` based on occurrence of newline bytes. If we encounter:
+            // 1. b'\r' then increment row count by 1.
+            // 2.a. If there was no b'\r' encountered before b'\n' then increment row count by 1.
+            // 2.b. If there was no b'\r' encountered before b'\n' then don't increment as b'\r\n' should be counted as 1 based on windows line ending pattern.
+           let (_, rows) = data.iter().fold((false, 0), |(follows_cr, rows), b| {
+                match (b, follows_cr) {
+                    // When there's a '\r', add a row
+                    (b'\r', _) => (true, rows + 1),
+                    // When there's a '\n' not after '\r', add a row
+                    (b'\n', false) => (false, rows + 1),
+                    // When there's '\n' immediately following '\r' without adding a row
+                    (b'\n', true) => (false, rows),
+                    _ => (false, rows),
+                }
+            });
+
+            rows + 1
+        } else {
+            1
+        }
+    }
+
+    #[cfg(feature = "lazy-source-location")]
+    fn prev_newline_offset(&self) -> usize {
+        // Update previous newline offset for location metadata
+        let available_range = ..self.prev_position;
+        let data  = &self.stream_bytes()[available_range];
+        if !data.is_empty() {
+            // Calculate `prev_newline_offset` based on the index/offset value of the last seen newline byte.
+            // Adding 1 to the index because we want to include everything after the newline -
+            // if newline is at index 5, we want to start counting from index 6 (5 + 1).
+            let prev_newline_offset = data.iter().enumerate().fold(0, |offset, (i, b)| {
+                match b {
+                    // When there's a '\r' or '\n', update the offset as this newline offset value
+                    b'\r' | b'\n' => i + 1,
+                  _ => offset,
+                }
+            });
+
+            prev_newline_offset
+        } else {
+            1
+        }
+    }
+
+    #[cfg(feature = "lazy-source-location")]
+    fn newlines(&self) -> &[usize] {
+        // This will act as a no-op, as for `IonSlice` we don't need to rely on newlines.
+        // It is only used by `IonStream`, to help keep track of all past newlines offsets.
+        &[]
+    }
```

Comment:
This source location implementation for `IonSlice` seems like it's going to be pretty inefficient. Both `row()` and `prev_newline_offset()` iterate through the entire source data every time they are called.

Instead, consider adding a `newlines: OnceCell<Vec<usize>>` (or `source_location_state: OnceCell<SourceLocationState>`) to `IonSlice` and then storing all of the newline locations there. Yes, it requires some auxiliary storage, but then all of the newline positions can be found once, and reused for every call to `row()` and `prev_newline_offset()`.

Reply:
I think this implementation will still be faster then storing location information. Since this `row` and `prev_newline_offset` are only called if user asks for location metadata i.e when `LazyValue#location()` is called. And I think since now it is only called when user asks for location its not that much of an overhead, unlike storing newlines and keeping it updated with each consumption might be expensive.

Reply:
Is it okay to iterate over this later, I can create an issue for it. As I am not sure If this is really going to improve performance I would need to run benchmarks for this particular suggestion and also because I think the current solution/lazy evaluation is faster.